datatypes/src/raster/signed_grid_dimension.rs

```diff
@@ -0,0 +1,754 @@
+/// An example for this is a grid wich is anchored at a geographical coordinate (0,0) but the upper left pixel is at (-180, 90).
+pub trait OffsetDimension: Sized {
+    type SignedIndexType: SignedGridIndex<Self>;
+    type DimensionType: GridDimension<IndexType = Self::UnsignedIndexType>;
+    type UnsignedIndexType: GridIndex<Self::DimensionType>;
+
+    /// The number of axis of the dimension.
+    const NDIM: usize;
+
+    /// The number of axis of the dimension.
+    fn number_of_dimensions(&self) -> usize {
+        Self::NDIM
+    }
+
+    /// The offsets for the dimension axis
+    fn offsets_as_slice(&self) -> &[SignedIdx];
+
+    /// This returns the offsets as index which are the min values of each axis. The values are INSIDE the bouds of the axis.
+    fn offsets_as_index(&self) -> Self::SignedIndexType;
+
+    /// This returns the max index of each axis. The values are INSIDE the bouds of the axis.
+    fn offsets_max_index(&self) -> Self::SignedIndexType;
+
+    // dimension
+    fn grid_dimension(&self) -> Self::DimensionType;
+
+    /// converts an offset index to a zero based by shifting the origin of each axis to zero.
+    fn offset_index_to_zero_based_unchecked(
+        &self,
+        index: &Self::SignedIndexType,
+    ) -> Self::UnsignedIndexType;
+
+    /// converts an offset index to a zero based by shifting the origin of each axis to zero
+    fn offset_index_to_zero_based(
+        &self,
+        index: &Self::SignedIndexType,
+    ) -> Result<Self::UnsignedIndexType> {
+        ensure!(
+            self.offset_index_inside_dimension(index),
+            error::GridSignedIndexOutOfBounds {
+                index: Vec::from(index.as_index_array().as_ref()),
+                dimension: Vec::from(self.grid_dimension().slice()),
+                offsets: Vec::from(self.offsets_as_slice())
+            }
+        );
+        Ok(self.offset_index_to_zero_based_unchecked(index))
+    }
+
+    /// converts an offset index directly into a zero based linear space index
+    fn offset_index_to_linear_space_index(&self, index: &Self::SignedIndexType) -> Result<usize> {
+        let zero_based = self.offset_index_to_zero_based(index)?;
+        Ok(self
+            .grid_dimension()
+            .linear_space_index_unchecked(&zero_based))
+    }
+
+    /// converts an offset index directly into a zero based linear space index
+    fn offset_index_to_linear_space_index_unchecked(&self, index: &Self::SignedIndexType) -> usize {
+        let zero_based = self.offset_index_to_zero_based_unchecked(index);
+        self.grid_dimension()
+            .linear_space_index_unchecked(&zero_based)
+    }
+
+    /// Check if a dimension contains a given index
+    fn offset_index_inside_dimension(&self, index: &Self::SignedIndexType) -> bool;
+
+    fn intersection(&self, other: &Self) -> Option<Self>
+    where
+        Self: Sized;
+}
+pub trait SignedGridIndex<S>
```

Comment:
why do we need two kinds of indexes? can't all indexes just be signed?

Reply:
we can do this. however it is not natural to have a signed index for a grid where no negative values are allowed.

Reply:
indexing now only uses isize